src/buffer/buffer_stm_test.ml

```diff
@@ -0,0 +1,123 @@
+open QCheck
+open Util
+
+(** parallel STM tests of Buffer *)
+
+(* port from the QCSTM example *)
+module BConf =
+struct
+  type cmd =
+    | Contents
+    (* To_bytes | Sub | Blit *)
+    | Nth of int
+    | Length
+    | Clear
+    | Reset
+    | Add_char of char
+    (* Add_utf8_uchar | Add_utf_16le_uchar | Add_utf_16be_uchar *)
+    | Add_string of string
+    (*| Add_bytes of bytes*)
+    | Truncate of int
+  [@@deriving show { with_path = false }]
+
+  type state = char list (* in reverse *)
+  type sut = Buffer.t
+
+  let arb_cmd s =
+    QCheck.make ~print:show_cmd
+      (Gen.oneof [Gen.return Contents;
+                  Gen.map (fun i -> Nth i) Gen.small_nat;
+                  Gen.return Length;
+                  Gen.return Clear;
+                  Gen.return Reset;
+                  Gen.map (fun c -> Add_char c) Gen.char;
+                  Gen.map (fun s -> Add_string s) (Gen.string);
+                  Gen.map (fun i -> Truncate i) (let len = List.length s in
+                                                 if len = 0
+                                                 then Gen.return 0
+                                                 else Gen.int_bound (len - 1));
+                 ])
+  
+  let init_state  = []
+
+  let rev_explode s =
+    let chars = ref [] in
+    String.iter (fun c -> chars := c::!chars) s;
+    !chars
+
+  let explode s = List.rev (rev_explode s)
+  
+  (* changed *)
+  let next_state c s = match c with
+    | Contents -> s
+    | Nth _ -> s
+    | Length -> s
+    | Clear -> []
+    | Reset -> []
+    | Add_char ch -> ch::s
+    | Add_string str -> (rev_explode str)@s (*s@(explode str)*)
+    | Truncate i ->
+      let rec trunc buf n = match buf,n with
+        | [],0 -> []
+        | [],_ -> []
+        | _::_,0 -> []
+        | c::cs,_ -> c::trunc cs (n-1) in
+      List.rev (trunc (List.rev s) i)
+  
+  let init_sut () = Buffer.create 16
+  let cleanup b   = Buffer.reset b
+
+  let precond c s = match c with
+    | Truncate i -> i >= 0 && i <= List.length s
```

Comment:
Just a word of warning: `precond` is ignored in parallel mode currently #38.
I hope to fix it shortly though, so it is fine to just keep the code as above.